components/candle/progressionService.ts

```diff
@@ -0,0 +1,238 @@
+/*
+ *     The Peacock Project - a HITMAN server replacement.
+ *     Copyright (C) 2021-2023 The Peacock Project Team
+ *
+ *     This program is free software: you can redistribute it and/or modify
+ *     it under the terms of the GNU Affero General Public License as published by
+ *     the Free Software Foundation, either version 3 of the License, or
+ *     (at your option) any later version.
+ *
+ *     This program is distributed in the hope that it will be useful,
+ *     but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *     GNU Affero General Public License for more details.
+ *
+ *     You should have received a copy of the GNU Affero General Public License
+ *     along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+import { getSubLocationByName } from "../contracts/dataGen"
+import { controller } from "../controller"
+import { grantDrops, getDataForUnlockables } from "../inventory"
+import type {
+    ContractSession,
+    UserProfile,
+    Unlockable,
+    GameVersion,
+} from "../types/types"
+import {
+    DEFAULT_MASTERY_MAXLEVEL,
+    clampValue,
+    xpRequiredForLevel,
+    xpRequiredForEvergreenLevel,
+    levelForXp,
+    evergreenLevelForXp,
+    getMaxProfileLevel,
+} from "../utils"
+import { writeUserData } from "../databaseHandler"
+
+export class ProgressionService {
+    //NOTE: Official will always grant XP to both Location Mastery and the Player Profile
+    grantProfileProgression(
+        actionXp = 0,
+        masteryXp = 0,
```

Comment:
Set these to "number", not a default value. That makes it the responsibility of something else not to pass undefined values and expect things to work.